src/log.cpp

```diff
@@ -218,33 +220,58 @@ std::ostream &logger::operator()(const log_domain& domain, bool show_names, bool
 			std::cerr << ss.str() << std::endl;
 			strict_threw_ = true;
 		}
-		std::ostream& stream = output();
+		log_in_progress stream = output();
 		if(do_indent) {
-			for(int i = 0; i != indent; ++i)
-				stream << "  ";
-			}
+			stream.set_indent(indent);
+		}
 		if (timestamp) {
-			if(precise_timestamp) {
-				print_precise_timestamp(stream);
-			} else {
-				stream << get_timestamp(std::time(nullptr));
-			}
+			stream.enable_timestamp();
 		}
 		if (show_names) {
-			stream << name_ << ' ' << domain.domain_->first << ": ";
+			stream.set_prefix(formatter() << name_ << ' ' << domain.domain_->first << ": ");
 		}
 		return stream;
 	}
 }
 
-void scope_logger::do_log_entry(const log_domain& domain, const std::string& str) noexcept
+log_in_progress::log_in_progress(std::ostream& stream)
+	: stream_(stream)
+{}
+
+void log_in_progress::operator|(formatter&& message)
+{
+	std::lock_guard<std::mutex> lock(log_mutex);
+	for(int i = 0; i < indent; ++i)
+		stream_ << "  ";
+	if(timestamp_) {
+		if(precise_timestamp) {
+			print_precise_timestamp(stream_);
+		} else {
+			stream_ << get_timestamp(std::time(nullptr));
+		}
+	}
+	stream_ << prefix_ << message.str();
+}
+
+void log_in_progress::set_indent(int level) {
+	indent_ = level;
+}
+
+void log_in_progress::enable_timestamp() {
+	timestamp_ = true;
+}
+
+void log_in_progress::set_prefix(const std::string& prefix) {
+	prefix_ = prefix;
+}
+
+void scope_logger::do_log_entry(const std::string& str) noexcept
```

Comment:
`noexcept` on a function that's doing I/O and manipulating strings? It was there before, but I think that indicates a bug in the existing code.

Reply:
Yeah, I'm not really sure. I agree that it seems suspect, but it's probably out-of-scope for this PR in any case.